File: src/products/getCategory.ts
import { DynamoDB } from 'aws-sdk';
import { HttpRequest, HttpResponse } from '../http'

const ddb = new DynamoDB.DocumentClient({ region: "eu-central-1" })

export async function handler(event: HttpRequest): Promise<HttpResponse> {

  const params = {
    TableName: process.env.PRODUCTS_TABLE!,
    KeyConditionExpression: "category = :category",
    ExpressionAttributeValues: {
      ":category": event.pathParameters.category
    }
  }

  try {
    const data = await ddb.query(params).promise()
    if (data) {
      return {
        statusCode: 200,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Credentials': true,
        },
        body: JSON.stringify(data.Items)
      }
    }
    return {
      statusCode: 404,
      body: "Category not found"
    }
  } catch (err) {
    return {
      statusCode: err.statusCode || 500,
      body: "Internal Server Error Oops" + err
    }
  }
}

File: dynamo_go/lambda/products/getCategory/getCategory.go
package main

import (
	"encoding/json"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
	"os"
	"webshop/internal/apigw"
	"webshop/internal/forms"
)

func Handler(event events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
	ddb := dynamodb.New(session.New())

	out, err := ddb.Query(&dynamodb.QueryInput{
		TableName:              aws.String(os.Getenv("PRODUCTS_TABLE")), //aws.String macht wert zu einem pointer
		KeyConditionExpression: aws.String("category = :category"),
		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
			":category": {
				S: aws.String(event.PathParameters["category"]),
			},
		},
	})

	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	if len(out.Items) > 0 {
		var products []*forms.Product
		for _, i := range out.Items {
			var product forms.Product
			if err := dynamodbattribute.UnmarshalMap(i, &product); err != nil {
				return apigw.CreateResponse(500, err.Error()), err
			}
			products = append(products, &product)
		}

		jsonStr, err := json.Marshal(products)
		if err != nil {
			return apigw.CreateResponse(500, err.Error()), err
		}

		return apigw.CreateResponse(200, string(jsonStr)), err
	}
	return apigw.CreateResponse(200, "[]"), err
}

func main() {
	lambda.Start(Handler)
}
